druid-shell/src/platform/x11/window.rs

```diff
@@ -224,6 +224,14 @@ impl WindowBuilder {
                 | EventMask::PointerMotion,
         );
 
+        // Conversion factor for scale (units = mm/pixels)
+        const FACTOR: f64 = 25.4 / 96.0;
+        let scale = Scale::new(
+            FACTOR * screen.width_in_pixels as f64 / screen.width_in_millimeters as f64,
+            FACTOR * screen.height_in_pixels as f64 / screen.height_in_millimeters as f64,
+        );
```

Comment:
I wouldn't rely on this information, for two reasons. First, I have no idea where it comes from and whether its reliable. On my system, xrandr and xdpyinfo report two different physical sizes for the same screen. Second, I think it's a bad idea to default to non-integer scaling. I'd argue for the xft-based solution that @crsaracco [linked](https://github.com/linebender/druid/issues/939#issuecomment-649853164) in #939.

Reply:
Thanks, I'll try implementing the xft method.

> Second, I think it's a bad idea to default to non-integer scaling.

You're joking, right? 1080p screens are common across 13" - 17" laptops now (165 - 130 DPI). 27" 1440p screens are also common (109 DPI). Besides which, many older users like slightly larger text. This is a *huge* demand for scale factors between 1 and 2.

Reply:
I wasn't actually joking, but I'm definitely happy to discuss it further. To be clear, I'm not arguing against *supporting* fractional scaling, just about defaulting to it. My main reason is that it causes anti-aliasing artifacts (like #1088).

> Besides which, many older users like slightly larger text.

Is there a reason to prefer scaling everything over just increasing the font size? (This is a genuine question -- I like *tiny* text so I'm very much not in the target audience)

Reply:
> that it causes anti-aliasing artifacts

Which is why I took a rather different approach to scaling with KAS (e.g. [here](https://github.com/kas-gui/kas/blob/master/kas-theme/src/dim.rs#L18)). But that's off topic...

> Is there a reason to prefer scaling everything over just increasing the font size?

Good question. Actually, I don't know much about how Druid likes to handle scaling, but at least the GTK backend currently is happy to use fractional scaling. This whole topic should probably get a new issue since this PR is just about fixing the currently-missing scaling support on X11 and broken support on GTK.

Reply:
> First, I have no idea where it comes from and whether its reliable. On my system, xrandr and xdpyinfo report two different physical sizes for the same screen. 

The data reported by `xdpyinfo` / the connection setup will always report 96 DPI. IIRC Keith Packard wrote that in a blog post somewhen, but I am not sure. But this can definitely not report data per RandR-output and there is also no way for the X11 server to update the information (well, RandR has something about that, but then you can just use the "real" info from RandR).

Reply:
I'm happy to see this merged as an improvement, and then revisit the larger question of how we want to 'properly' do scaling? I'm not spending much time on linux (although I'm happy to play around in a vm as needed) so I trust @jneem and @dhardy to get this merged when there's consensus, and I'm happy to break any ties if needed. 👼

Reply:
Perhaps we should also refer to the work of our peers: https://docs.rs/winit/0.23.0/winit/dpi/index.html#how-is-the-scale-factor-calculated

I'm in two minds whether it's worth bothering much here.

Reply:
Winit uses [Xlib::open](https://docs.rs/x11-dl/2.18.5/x11_dl/xlib/struct.Xlib.html#method.open) to access `Xft.dpi`. The `RustAudio` code uses the `x11` crate. I think we can't access X resources with `x11rb`, so an extra dependency would be necessary?

Reply:
At least from a C point of view: You are already linking against `libcairo.so` even when using x11rb. `libcairo.so` links against `libX11`. So your extra dependency would only be a Rust wrapper for `libX11`.

Also: Due to this issue and https://github.com/psychon/x11rb/issues/550, I noticed that Xresource handling is missing in x11rb and opened https://github.com/psychon/x11rb/issues/551 about that. It might be that the next release of x11rb will provide a way to query `Xft.dpi`. Technically, it is really easy. It's just that I'd have to hand-write a parser for the necessary file format (easy but annoying) and be reasonable sure that it behaves as expected (this seems to be the hard case - Xresources seem to be quite powerful and I do not really understand all the rules (yet)).
X11-the-protocol-wise, Xresources are just a property on the root window containing a string, plus some fallbacks in case there is no such properties. Thus, only the parsing and the actual query matching are necessary.

Reply:
I certainly do not claim to know what I am doing, but I just added an example to https://github.com/psychon/x11rb/issues/550#issuecomment-729025335 that draws a 2 cm line by querying information from RandR. It works for me, so it must obviously work everywhere. ;-)
This does *not* do anything with Xft configuration and has zero user configurability. It just detects values and uses them.
Also, I screwed up the calculation of DPI three times and then gave up. Thus, the code works in dots per cm instead. This also makes it a lot easier to calculate the length of a 2 cm line. This should just be a matter of adding the right scaling factor.

Also: Urgh for X11 (and thus x11rb) mixing u16, i16 and u32 in the involved APIs.